InformationScripting/src/queries/LinearQuery.h

```diff
@@ -0,0 +1,47 @@
+/***********************************************************************************************************************
+**
+** Copyright (c) 2011, 2015 ETH Zurich
+** All rights reserved.
+**
+** Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+** following conditions are met:
+**
+**    * Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+**      disclaimer.
+**    * Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+**      following disclaimer in the documentation and/or other materials provided with the distribution.
+**    * Neither the name of the ETH Zurich nor the names of its contributors may be used to endorse or promote products
+**      derived from this software without specific prior written permission.
+**
+**
+** THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+** INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+** DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+** SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+** SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+** WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+** OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+**
+***********************************************************************************************************************/
+
+#pragma once
+
+#include "../informationscripting_api.h"
+
+#include "Query.h"
+
+namespace InformationScripting {
+
+class INFORMATIONSCRIPTING_API LinearQuery : public Query
+{
+	public:
+		virtual QList<TupleSet> execute(QList<TupleSet> input) override final;
+		virtual TupleSet execute(TupleSet) = 0;
+
+	protected:
+		template <class QueryClass>
+		using LinearExecuteFunction = std::function<TupleSet (QueryClass*, TupleSet)>;
```

Comment:
OK, I looked and couldn't find a good way of doing this, although I have a strong feeling I'm missing something, and there should be a better way. In any case, you can leave the mechanism as it is.

But I still think you should move this declaration (and create two copies of it), into the most-derived classes that use it. In a sense this declaration is an implementation detail of those classes. I might make a different class where I choose to work with a big switch statement in the `execute` function, or do something else entirely, and this declaration will be of no use to me. Even though you will create two copies of it, by moving it up, I think that's fine, since those two copies are in a sense accidental. We just happen to implement the two different sources in a similar way.